File: sdk/rust/zhtw/src/builder.rs
use crate::config::AmbiguityMode;
use crate::{Config, Result, Source};

#[derive(Debug, Default)]
pub struct Builder {
    config: Config,
}

impl Builder {
    pub fn sources<I: IntoIterator<Item = Source>>(mut self, sources: I) -> Self {
        self.config.sources = sources.into_iter().collect();
        self
    }

    pub fn custom_dict<I, K, V>(mut self, dict: I) -> Self
    where
        I: IntoIterator<Item = (K, V)>,
        K: Into<String>,
        V: Into<String>,
    {
        self.config.custom_dict = dict
            .into_iter()
            .map(|(k, v)| (k.into(), v.into()))
            .collect();
        self
    }

    pub fn ambiguity_mode(mut self, mode: AmbiguityMode) -> Self {
        self.config.ambiguity_mode = mode;
        self
    }

    pub fn build(self) -> Result<crate::Converter> {
        crate::Converter::new(self.config)
    }
}

File: sdk/go/zhtw/builder.go
package zhtw

// Builder constructs a custom Converter with non-default settings.
type Builder struct {
	sources       []Source
	customDict    map[string]string
	ambiguityMode AmbiguityMode
}

// NewBuilder returns a Builder with default settings (Cn+Hk, Strict).
func NewBuilder() *Builder {
	return &Builder{
		sources:       []Source{SourceCn, SourceHk},
		ambiguityMode: AmbiguityStrict,
	}
}

// Sources sets the dictionary sources. Default: [SourceCn, SourceHk].
func (b *Builder) Sources(sources ...Source) *Builder {
	b.sources = sources
	return b
}

// CustomDict sets user overrides that take priority over built-in terms.
func (b *Builder) CustomDict(dict map[string]string) *Builder {
	b.customDict = dict
	return b
}

// SetAmbiguityMode sets the ambiguity handling mode. Default: AmbiguityStrict.
func (b *Builder) SetAmbiguityMode(mode AmbiguityMode) *Builder {
	b.ambiguityMode = mode
	return b
}

// Build creates the Converter. Returns an error if sources is empty.
func (b *Builder) Build() (*Converter, error) {
	data := getParsedData()
	return buildConverter(data, b.sources, b.customDict, b.ambiguityMode)
}
